addons/mail/models/mail_thread.py

```diff
@@ -2304,6 +2287,31 @@ def send_notifications():
 
         return True
 
+    def get_template_generator(self, partners_data, msg_vals, message, model_description):
+        model = msg_vals.get('model') if msg_vals else message.model
+        email_layout_xmlid = msg_vals.get('email_layout_xmlid') if msg_vals else message.email_layout_xmlid
+        template_xmlid = email_layout_xmlid if email_layout_xmlid else 'mail.message_notification_email'
+        for (lang, partners_data_batch) in self.group_partners_data_by_lang(partners_data):
+            self = self.with_context(lang=lang)
+            model_name = model_description or (self.env['ir.model']._get(model).display_name if model else False) # one query for display name
+            groups = self._notify_classify_recipients(partners_data_batch, model_name, msg_vals=msg_vals)
+            if not groups:
+                continue
+            template_values = self._notify_prepare_template_context(message, msg_vals, model_description=model_description) # 10 queries
+            try:
+                base_template = self.env.ref(template_xmlid, raise_if_not_found=True) # 1 query
+            except ValueError:
+                _logger.warning('QWeb template %s not found when sending notification emails. Sending without layouting.', template_xmlid)
+                base_template = False
+            for recipients_group_data in groups:
+                yield (base_template, template_values, recipients_group_data)
+
+    def group_partners_data_by_lang(self, partners_data):
```

Comment:
It would be cool to have a generic function to group the tuples in a dictionary like:

```python3
def group_by(iterable, key):
    batches = {}
    for current in iterable:
        batches.setdefault(key(current), []).append(current)
    return batches

print(group_by([(1, 2), (4, 5), (1, 3)], key=lambda elem: elem[0])) # print {1: [(1, 2), (1, 3)], 4: [(4, 5)]}
```

In the example, we would have something like this:

```
group_by(partners_data, key=lambda current: current.get('lang') or self.env.lang)
```

Is there any file where I can add this helper function ?